### PR TITLE
Support new gnmi config interface in telemetry container.

### PR DESCRIPTION
**Pull request by ganglyu**
Signed-off-by: Gang Lv ganglv@microsoft.com

#### Why I did it
Update telemetry container to support new gnmi config interface.

#### How I did it
Use new target to support mixed schema, and we can enable/disable telemetry and gnmi config interface.

#### How to verify it
Build target and run unit test.

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111

### CODE REVIEW

File: azure-pipelines.yml [hunk 1]
```diff
@@ -40,7 +40,7 @@ stages:
       DIFF_COVER_WORKING_DIRECTORY: $(System.DefaultWorkingDirectory)/sonic-gnmi
 
     container:
-      image: sonicdev-microsoft.azurecr.io:443/sonic-slave-buster:latest
+      image: sonicdev-microsoft.azurecr.io:443/sonic-slave-bullseye:latest
```
Comment (qiluo-msft) [inline azure-pipelines.yml:43]: >bullseye [](http://example.com/codeflow?start=59&length=8)

Let's separate irrelevant changes such as "bullseye" to another PR and focus "native mode gnmi service" in this PR.

If you mix, it's challenge to revert anything in case.

File: Makefile [hunk 2]
```diff
@@ -18,7 +18,15 @@ TEST_FILES=$(wildcard *_test.go)
 TELEMETRY_TEST_DIR = build/tests/gnmi_server
 TELEMETRY_TEST_BIN = $(TELEMETRY_TEST_DIR)/server.test
 ifeq ($(ENABLE_TRANSLIB_WRITE),y)
-BLD_FLAGS := -tags gnmi_translib_write
+  ifeq ($(ENABLE_NATIVE_WRITE),y)
```
Comment (qiluo-msft) [inline Makefile:21]: >ENABLE_NATIVE_WRITE [](http://example.com/codeflow?start=10&length=19)

You may check each ENABLE_XXX flag, and append to string, and finally set BLD_FLAGS by the string. Current nested if-else is not scalable for future extra flags.

File: Makefile [hunk 3]
```diff
@@ -41,38 +49,48 @@ $(GO_DEPS): go.mod $(PATCHES)
 	patch -d vendor -p0 < patches/gnmi_cli.all.patch
 	patch -d vendor -p0 < patches/gnmi_set.patch
 	patch -d vendor -p0 < patches/gnmi_get.patch
+	patch -d vendor -p0 < patches/path.patch
```
Comment (qiluo-msft) [inline Makefile:52]: >path.patch [](http://example.com/codeflow?start=31&length=10)

The patch filename is too general.

File: Makefile [hunk 4]
```diff
@@ -41,38 +49,48 @@ $(GO_DEPS): go.mod $(PATCHES)
...
 	$(GO) install -mod=vendor $(BLD_FLAGS) github.com/sonic-net/sonic-gnmi/telemetry
 	$(GO) install -mod=vendor $(BLD_FLAGS) github.com/sonic-net/sonic-gnmi/dialout/dialout_client_cli
 	$(GO) install -mod=vendor github.com/jipanyang/gnxi/gnmi_get
 	$(GO) install -mod=vendor github.com/jipanyang/gnxi/gnmi_set
 	$(GO) install -mod=vendor github.com/openconfig/gnmi/cmd/gnmi_cli
 	$(GO) install -mod=vendor github.com/sonic-net/sonic-gnmi/gnoi_client
+	$(GO) install -mod=vendor github.com/sonic-net/sonic-gnmi/gnmi_dump
 endif
 
+# TODO: Create a new repo for this lib, sonic-restapi and sonic-gnmi can share this lib
```
Comment (qiluo-msft) [inline Makefile:79]: >TODO [](http://example.com/codeflow?start=2&length=4)

Instead of todo, let's do it now?

File: gnmi_server/gnoi.go [hunk 5]
```diff
@@ -2,27 +2,59 @@ package gnmi
...
 	spb_jwt "github.com/sonic-net/sonic-gnmi/proto/gnoi/jwt"
+	"github.com/sonic-net/sonic-gnmi/common_utils"
 	"google.golang.org/grpc/status"
 	"google.golang.org/grpc/codes"
 	"os/user"
 	"encoding/json"
 	jwt "github.com/dgrijalva/jwt-go"
 )
 
+func RebootSystem(fileName string) error {
```
Comment (qiluo-msft) [inline gnmi_server/gnoi.go:23]: >RebootSystem [](http://example.com/codeflow?start=5&length=12)

Is it "reboot" or "config reload"?

File: gnmi_server/gnoi.go [hunk 6]
```diff
@@ -2,27 +2,59 @@ package gnmi
...
+	if err != nil {
+		return err
+	}
+	err = sc.ConfigReload(fileName)
+	return err
+}
+
 func (srv *Server) Reboot(ctx context.Context, req *gnoi_system_pb.RebootRequest) (*gnoi_system_pb.RebootResponse, error) {
+	fileName := common_utils.GNMI_WORK_PATH + "/config_db.json.tmp"
+	//return nil, status.Errorf(codes.Unimplemented, "hit test point")
```
Comment (qiluo-msft) [inline gnmi_server/gnoi.go:35]: >//return nil, status.Errorf(codes.Unimplemented, "hit test point") [](http://example.com/codeflow?start=1&length=66)

Remove test code, not just commented.

File: gnmi_server/server.go [hunk 7]
```diff
@@ -274,26 +284,32 @@ func (s *Server) checkEncodingAndModel(encoding gnmipb.Encoding, models []*gnmip
 
 // Get implements the Get RPC in gNMI spec.
 func (s *Server) Get(ctx context.Context, req *gnmipb.GetRequest) (*gnmipb.GetResponse, error) {
+	common_utils.IncCounter("GNMI get")
```
Comment (qiluo-msft) [inline gnmi_server/server.go:287]: >IncCounter [](http://example.com/codeflow?start=14&length=10)

The counter and dump are quite useful new features, not directly tie to the native mode gnmi service feature. Could you separate them into separate PR(s)?

**Comment by qiluo-msft** [comment]
> func (srv *Server) CancelReboot(ctx context.Context, req *gnoi_system_pb.CancelRebootRequest) (*gnoi_system_pb.CancelRebootResponse, error) {

Is it a placeholder function (with TODO comment)? or how did you cancel a reboot?

---
Refers to: gnmi_server/gnoi.go:66 in 474b344. [](commit_id = 474b34406f11c3780524ebe7cbff384638a86bf6, deletion_comment = False)

File: patches/path.patch [hunk 8]
```diff
@@ -0,0 +1,37 @@
+--- ./github.com/jipanyang/gnxi/utils/xpath/path.go
```
Comment (qiluo-msft) [inline patches/path.patch:1]: >---  [](http://example.com/codeflow?start=0&length=4)

Could you add some file level comment to explain why a patch is needed, and how do you achieve it?
ref: [https://stackoverflow.com/a/20911434/2514803](https://stackoverflow.com/a/20911434/2514803)

File: sonic_data_client/events_client.go [hunk 9]
```diff
@@ -275,7 +275,7 @@ func get_events(evtc *EventClient) {
     for {
 
         rc, evt := C_recv_evt(evtc.subs_handle)
-        log.V(7).Infof("C.event_receive_wrap rc=%d evt:%s", rc, (*C.char)(str_ptr))
+        log.V(7).Infof("C.event_receive_wrap rc=%d evt:%v", rc, (*C.char)(str_ptr))
```
Comment (qiluo-msft) [inline sonic_data_client/events_client.go:278]: >v [](http://example.com/codeflow?start=56&length=1)

For valid bug fix, separate into a separate PR.

File: sonic_data_client/transl_data_client.go [hunk 10]
```diff
@@ -204,7 +204,7 @@ func (c *TranslClient) StreamRun(q *queue.PriorityQueue, stop chan struct{}, w *
 				return
 			}
 		default:
-			log.V(1).Infof("Bad Subscription Mode for client %s ", c)
```
Comment (qiluo-msft) [inline sonic_data_client/transl_data_client.go:207]: >s [](http://example.com/codeflow?start=53&length=1)

The same

File: gnmi_server/gnoi.go (hunk 5 above)
Comment (ganglyu) [reply gnmi_server/gnoi.go:23]: Use reboot to implement "config reload", and please refer to: https://github.com/openconfig/gnoi/blob/main/system/system.proto#L131

File: libcswsscommon/include/capi/dbconnector.h [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+#include <stdbool.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef void *db_connector_t;
+
+// DBConnector::DBConnector(int db, std::string hostname, int port, unsigned int timeout)
+db_connector_t db_connector_new(int db, const char *hostname, int port, unsigned int timeout);
```
Comment (qiluo-msft) [inline libcswsscommon/include/capi/dbconnector.h:14]: >db_connector_new [](http://example.com/codeflow?start=15&length=16)

Did you try SWIG?
[http://zacg.github.io/blog/2013/06/06/calling-c-plus-plus-code-from-go-with-swig/](http://zacg.github.io/blog/2013/06/06/calling-c-plus-plus-code-from-go-with-swig/)

File: gnmi_server/server.go [hunk 12]
```diff
@@ -274,6 +277,26 @@ func (s *Server) checkEncodingAndModel(encoding gnmipb.Encoding, models []*gnmip
...
+			}
+		} else if origin != path.Origin {
+			return "", status.Error(codes.Unimplemented, "Origin conflict in path")
+		}
+	}
+	return origin, nil
+}
+
+func IsNativeOrigin(origin string) bool {
+	return origin == "sonic-db" || origin == "sonic-yang"
```
Comment (qiluo-msft) [inline gnmi_server/server.go:297]: >"sonic-db" [](http://example.com/codeflow?start=18&length=10)

Origin `sonic-db` is named as `native` since the gnmi client directly manipulate raw ConfigDB. `sonic-yang` is not considered `native` because there will be a translation between sonic-yang payload to ConfigDB schema.

File: gnmi_server/server.go [hunk 13]
```diff
@@ -274,6 +277,26 @@ func (s *Server) checkEncodingAndModel(encoding gnmipb.Encoding, models []*gnmip
 	return nil
 }
 
+func ParseOrigin(origin string, paths []*gnmipb.Path) (string, error) {
+	if len(paths) == 0 {
+		return origin, nil
+	}
+	for i, path := range paths {
+		if origin == "" {
```
Comment (qiluo-msft) [inline gnmi_server/server.go:285]: >origin [](http://example.com/codeflow?start=5&length=6)

Why treat empty input `origin` differently?

File: gnmi_server/server.go [hunk 14]
```diff
@@ -274,6 +277,26 @@ func (s *Server) checkEncodingAndModel(encoding gnmipb.Encoding, models []*gnmip
 	return nil
 }
 
+func ParseOrigin(origin string, paths []*gnmipb.Path) (string, error) {
```
Comment (qiluo-msft) [inline gnmi_server/server.go:280]: >origin [](http://example.com/codeflow?start=17&length=6)

why a parse function has an extra input parameter `origin`? Could you add a function level comment?

File: gnmi_server/server.go (hunk 13 above)
Comment (ganglyu) [reply gnmi_server/server.go:285]: I will fix this.

File: gnmi_server/server.go (hunk 14 above)
Comment (ganglyu) [reply gnmi_server/server.go:280]: I will fix this.

**Review by qiluo-msft — APPROVED** [review]
